README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/bottom-sheet.svg)](https://www.npmjs.com/package/@advanced-rest-client/bottom-sheet)

[![Build Status](https://travis-ci.com/advanced-rest-client/bottom-sheet.svg)](https://travis-ci.com/advanced-rest-client/bottom-sheet)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/bottom-sheet)


# &lt;bottom-sheet&gt;

Bottom sheets slide up from the bottom of the screen to reveal more content.

## Example:

```html
<bottom-sheet nopadding>
  <paper-item>Action #1</paper-item>
  <paper-item>Action #2</paper-item>
  <paper-item>Action #3</paper-item>
</bottom-sheet>
```

## Usage

### Installation
```
npm install --save @advanced-rest-client/bottom-sheet
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import './node_modules/@advanced-rest-client/bottom-sheet/bottom-sheet.js';
    </script>
  </head>
  <body>
    <bottom-sheet>
      <paper-item>Action #1</paper-item>
      <paper-item>Action #2</paper-item>
      <paper-item>Action #3</paper-item>
    </bottom-sheet>
    <paper-button>Open menu</paper-button>
    <script>
    {
      document.querySelector('paper-button').addEventListener('click', () => {
        document.querySelector('bottom-sheet').opened = true;
      });
    }
    </script>
  </body>
</html>
```

### In a LitElement

```js
import { LitElement, html } from 'lit-element';
import '@advanced-rest-client/bottom-sheet/bottom-sheet.js';

class SampleElement extends LitElement {
  static get properties() {
    return {
      menuOpened: { type: Boolean }
    }
  }
  render() {
    return html`
    <bottom-sheet .opened="${this.menuOpened}">
      <paper-item>Action #1</paper-item>
      <paper-item>Action #2</paper-item>
      <paper-item>Action #3</paper-item>
    </bottom-sheet>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

### In a Polymer 3 element

```js
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';
import '@advanced-rest-client/bottom-sheet/bottom-sheet.js';

class SampleElement extends PolymerElement {
  static get properties() {
    return {
      menuOpened: { type: Boolean, value: false }
    }
  }
  static get template() {
    return html`
    <bottom-sheet opened?="{{menuOpened}}">
      <paper-item>Action #1</paper-item>
      <paper-item>Action #2</paper-item>
      <paper-item>Action #3</paper-item>
    </bottom-sheet>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

## Development

```sh
git clone https://github.com/advanced-rest-client/bottom-sheet
cd bottom-sheet
npm install
```

### Running the demo locally

```sh
npm start
```

### Running the tests
```sh
polymer test
```


CHANGELOG.md
<a name="2.0.2"></a>
## [2.0.2](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...2.0.2) (2018-11-21)




<a name="2.0.1"></a>
## [2.0.1](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...2.0.1) (2018-11-21)




<a name="2.0.0"></a>
# [2.0.0](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...2.0.0) (2018-07-24)




<a name="1.1.2"></a>
## [1.1.2](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...1.1.2) (2017-10-03)




<a name="1.1.1"></a>
## [1.1.1](https://github.com/advanced-rest-client/bottom-sheet/compare/1.0.5...1.1.1) (2017-10-03)


### New

* Added support for scrollable content. ([ef97ff907f1c6f755c03f447498ab3717bad8e59](https://github.com/advanced-rest-client/bottom-sheet/commit/ef97ff907f1c6f755c03f447498ab3717bad8e59))

### Update

* Changed testing to new model. ([6b4b27f8e7422f5ea17daead83581b95448069f3](https://github.com/advanced-rest-client/bottom-sheet/commit/6b4b27f8e7422f5ea17daead83581b95448069f3))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/advanced-rest-client/bottom-sheet/compare/1.0.3...1.0.5) (2017-09-23)


### Fix

* Fixed "hidden" attribute styles ([c8738afaae88d7871d4d15a19b6f30d1799b86b1](https://github.com/advanced-rest-client/bottom-sheet/commit/c8738afaae88d7871d4d15a19b6f30d1799b86b1))

### Update

* Added new Travis configuration ([912d29dab2facfdb2b1b4ef19a7a8955cd2874ef](https://github.com/advanced-rest-client/bottom-sheet/commit/912d29dab2facfdb2b1b4ef19a7a8955cd2874ef))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/advanced-rest-client/bottom-sheet/compare/1.0.3...v1.0.4) (2016-11-20)


### Update

* Added new Travis configuration ([912d29dab2facfdb2b1b4ef19a7a8955cd2874ef](https://github.com/advanced-rest-client/bottom-sheet/commit/912d29dab2facfdb2b1b4ef19a7a8955cd2874ef))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/advanced-rest-client/bottom-sheet/compare/1.0.1...v1.0.3) (2016-10-06)


### New

* Added hero image ([3dbc9873a319397d383a159772bee5007f72f477](https://github.com/advanced-rest-client/bottom-sheet/commit/3dbc9873a319397d383a159772bee5007f72f477))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/advanced-rest-client/bottom-sheet/compare/1.0.1...v1.0.2) (2016-09-29)




<a name="1.0.1"></a>
## 1.0.1 (2016-09-12)


### Update

* Updated dependencies tree ([fe7e9efe9e5d3c22594a2c8d5ef7ea20fedcc0f3](https://github.com/advanced-rest-client/bottom-sheet/commit/fe7e9efe9e5d3c22594a2c8d5ef7ea20fedcc0f3))



# [3.0.0](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...3.0.0) (2019-05-31)


### Breaking

* Upgrading to polymer 3 ([92e9c2f7d4c6eb5696554855a4643e0e363bc78c](https://github.com/advanced-rest-client/bottom-sheet/commit/92e9c2f7d4c6eb5696554855a4643e0e363bc78c))

### Update

* Bumping version to stable ([40ce43dd82accbd445faf2388a94bf799e7d8d48](https://github.com/advanced-rest-client/bottom-sheet/commit/40ce43dd82accbd445faf2388a94bf799e7d8d48))
* Removing FF from local tests do to unknown WCT error ([88d35ba46704a68418825026a2091b9cbc18e3f2](https://github.com/advanced-rest-client/bottom-sheet/commit/88d35ba46704a68418825026a2091b9cbc18e3f2))
* Updating typpings ([e9d97b09c80b7a41543b9a3c7cf3b8bdb84a7ed6](https://github.com/advanced-rest-client/bottom-sheet/commit/e9d97b09c80b7a41543b9a3c7cf3b8bdb84a7ed6))
* Upgrading dependencies ([09b4fbf729169bf1b034f8021512d5324f65f625](https://github.com/advanced-rest-client/bottom-sheet/commit/09b4fbf729169bf1b034f8021512d5324f65f625))



# [3.1.0](https://github.com/advanced-rest-client/bottom-sheet/compare/1.1.1...3.1.0) (2019-07-23)


### Breaking

* Upgrading to polymer 3 ([92e9c2f7d4c6eb5696554855a4643e0e363bc78c](https://github.com/advanced-rest-client/bottom-sheet/commit/92e9c2f7d4c6eb5696554855a4643e0e363bc78c))

### docs

* restoring typpings ([24d9e797ca07a283a4f49a653361dba691441125](https://github.com/advanced-rest-client/bottom-sheet/commit/24d9e797ca07a283a4f49a653361dba691441125))
* updating readme file ([88ff8573446b7eedf4b8713c4c3e77309048d770](https://github.com/advanced-rest-client/bottom-sheet/commit/88ff8573446b7eedf4b8713c4c3e77309048d770))

### fix

* removing unwanted file ([09ff3376a146d25fe13300965c6fdb9eb3938a70](https://github.com/advanced-rest-client/bottom-sheet/commit/09ff3376a146d25fe13300965c6fdb9eb3938a70))

### refactor

* upgrading to LitElement ([11dc771508c3ea54b7ea2f1da5f47244c103f18d](https://github.com/advanced-rest-client/bottom-sheet/commit/11dc771508c3ea54b7ea2f1da5f47244c103f18d))

### Update

* Bumping version to stable ([40ce43dd82accbd445faf2388a94bf799e7d8d48](https://github.com/advanced-rest-client/bottom-sheet/commit/40ce43dd82accbd445faf2388a94bf799e7d8d48))
* Removing FF from local tests do to unknown WCT error ([88d35ba46704a68418825026a2091b9cbc18e3f2](https://github.com/advanced-rest-client/bottom-sheet/commit/88d35ba46704a68418825026a2091b9cbc18e3f2))
* Updating typpings ([e9d97b09c80b7a41543b9a3c7cf3b8bdb84a7ed6](https://github.com/advanced-rest-client/bottom-sheet/commit/e9d97b09c80b7a41543b9a3c7cf3b8bdb84a7ed6))
* Upgrading dependencies ([09b4fbf729169bf1b034f8021512d5324f65f625](https://github.com/advanced-rest-client/bottom-sheet/commit/09b4fbf729169bf1b034f8021512d5324f65f625))



<a name="3.2.0"></a>
## [3.2.0](https://github.com/advanced-rest-client/bottom-sheet/compare/3.0.0...3.2.0) (2020-07-19)

### Update

* upgrading dependencies to fix audit errors [a42558a](https://github.com/advanced-rest-client/bottom-sheet/commit/a42558ac37504dcc112880480c70e3254da0d86b) by Pawel Psztyc


### Documentation

* updating readme file [88ff857](https://github.com/advanced-rest-client/bottom-sheet/commit/88ff8573446b7eedf4b8713c4c3e77309048d770) by Pawel
* restoring typpings [24d9e79](https://github.com/advanced-rest-client/bottom-sheet/commit/24d9e797ca07a283a4f49a653361dba691441125) by Pawel


### Features


test/bottom-sheet.test.js
import { fixture, assert, aTimeout, oneEvent } from '@open-wc/testing';
import '../bottom-sheet.js';


/** @typedef {import('../index').BottomSheetElement} BottomSheetElement */

describe('<bottom-sheet>', () => {
  /**
   * @returns {Promise<BottomSheetElement>}
   */
  async function basicFixture() {
    return (fixture(`<bottom-sheet></bottom-sheet>`));
  }

  /**
   * @returns {Promise<BottomSheetElement>}
   */
  async function openedFixture() {
    return (fixture(`<bottom-sheet opened></bottom-sheet>`));
  }

  /**
   * @returns {Promise<BottomSheetElement>}
   */
  async function labelFixture() {
    return (fixture(`<bottom-sheet label="test"></bottom-sheet>`));
  }

  describe('Basics', () => {
    let sheet;

    it('is hidden', async () => {
      sheet = await basicFixture();
      assert.isFalse(sheet.opened);
    });

    it('is visible', async () => {
      sheet = await openedFixture();
      assert.isTrue(sheet.opened, '`opened` is true');
    });

    it('show() will open bottom-sheet', async () => {
      sheet = await basicFixture();
      sheet.open();
      assert.isTrue(sheet.opened, '`opened` is true');
    });

    it('hide() will close bottom-sheet', async () => {
      sheet = await openedFixture();
      sheet.close();
      assert.isFalse(sheet.opened, '`opened` is true');
    });

    it('bottom-sheet fires opened event', async () => {
      const element = await openedFixture();
      await oneEvent(element, 'iron-overlay-opened');
    });

    it('there is only 1 bottom-sheet opened', async () => {
      const sheet1 = await basicFixture();
      const sheet2 = await openedFixture();
      sheet2.open();
      sheet1.open();
      assert.isTrue(sheet1.opened, 'sheet1 is opened');
      assert.isFalse(sheet2.opened, 'sheet2 is not opened');
      sheet2.open();
      assert.isFalse(sheet1.opened, 'sheet1 is now not opened');
      assert.isTrue(sheet2.opened, 'sheet2 is now opened');
    });

    it('scrollTarget returns the target', async () => {
      sheet = await basicFixture();
      await aTimeout(0);
      assert.ok(sheet.scrollTarget);
    });

    it('_renderOpened() adds bottom-sheet-open class', async () => {
      sheet = await basicFixture();
      sheet._renderOpened();
      assert.isTrue(sheet.classList.contains('bottom-sheet-open'));
    });

    it('_renderClosed() removes bottom-sheet-open class', async () => {
      sheet = await basicFixture();
      sheet.className = 'bottom-sheet-open';
      sheet._renderClosed();
      assert.isFalse(sheet.classList.contains('bottom-sheet-open'));
    });
  });

  describe('a11y', () => {
    it('passes in a regular state', async () => {
      const element = await basicFixture();
      await assert.isAccessible(element);
    });

    it('passes in an opened state', async () => {
      const element = await openedFixture();
      await assert.isAccessible(element);
    });

    it('passes with a label', async () => {
      const element = await labelFixture();
      await assert.isAccessible(element);
    });
  });
});


demo/index.js
import { html, render } from 'lit-html';
import '@advanced-rest-client/arc-demo-helper/arc-demo-helper.js';
import '@polymer/paper-item/paper-item.js';
import '@polymer/paper-item/paper-icon-item.js';
import '@polymer/paper-button/paper-button.js';
import '@polymer/iron-icon/iron-icon.js';
import '@polymer/paper-toggle-button/paper-toggle-button.js';
import '../bottom-sheet.js';

class DemoPage {
  constructor() {
    this._open = this._open.bind(this);
    this._toggleTheme = this._toggleTheme.bind(this);
    this._toggleStyles = this._toggleStyles.bind(this);
  }

  _open(e) {
    e.target.previousElementSibling.opened = true;
  }

  _toggleTheme(e) {
    if (e.target.checked) {
      document.body.classList.add('dark');
    } else {
      document.body.classList.remove('dark');
    }
  }

  _toggleStyles(e) {
    if (e.target.checked) {
      document.body.classList.add('styled');
    } else {
      document.body.classList.remove('styled');
    }
  }

  render() {
    render(html`<div class="vertical-section-container centered">
      <h3>The bottom-sheet</h3>
      <bottom-sheet noPadding>
        <paper-item>Action #1</paper-item>
        <paper-item>Action #2</paper-item>
        <paper-item>Action #3</paper-item>
      </bottom-sheet>
      <paper-button raised @click="${this._open}">Open menu</paper-button>

      <h3>Open with ...</h3>
      <bottom-sheet class="center-bottom" label="Open with ..." noPadding>
        <paper-icon-item>
          <iron-icon src="inbox.png" slot="item-icon"></iron-icon>
          Inbox
        </paper-icon-item>
        <paper-icon-item>
          <iron-icon src="keep.png" slot="item-icon"></iron-icon>
          Keep
        </paper-icon-item>
        <paper-icon-item>
          <iron-icon src="hangouts.png" slot="item-icon"></iron-icon>
          Hangouts
        </paper-icon-item>
        <paper-icon-item>
          <iron-icon src="messenger.png" slot="item-icon"></iron-icon>
          Messenger
        </paper-icon-item>
        <paper-icon-item>
          <iron-icon src="gplus.png" slot="item-icon"></iron-icon>
          Google+
        </paper-icon-item>
      </bottom-sheet>
      <paper-button raised @click="${this._open}">Open with ...</paper-button>

      <h3>Scrollable content</h3>
      <bottom-sheet class="center-bottom" label="Images (might take time to load)">
        <img alt="i1" src="http://loremflickr.com/400/400/?random=0">
        <img alt="i2" src="http://loremflickr.com/400/400/?random=1">
        <img alt="i3" src="http://loremflickr.com/400/400/?random=2">
        <img alt="i4" src="http://loremflickr.com/400/400/?random=3">
        <img alt="i5" src="http://loremflickr.com/400/400/?random=4">
        <img alt="i6" src="http://loremflickr.com/400/400/?random=5">
        <img alt="i7" src="http://loremflickr.com/400/400/?random=6">
        <img alt="i8" src="http://loremflickr.com/400/400/?random=7">
        <img alt="i9" src="http://loremflickr.com/400/400/?random=8">
        <img alt="i10" src="http://loremflickr.com/400/400/?random=9">
      </bottom-sheet>
      <paper-button raised @click="${this._open}">Open</paper-button>
    </div>

    <div class="centered actions card">
      <h2>Options</h2>
      <paper-toggle-button @checked-changed="${this._toggleTheme}">Toggle dark theme</paper-toggle-button>
      <paper-toggle-button @checked-changed="${this._toggleStyles}" checked>Toggle styles</paper-toggle-button>
    </div>
    `, document.querySelector('#demo'));
  }
}
const instance = new DemoPage();
instance.render();


src/BottomSheetElement.d.ts
/**
@license
Copyright 2018 The Advanced REST Client authors
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, TemplateResult, CSSResult } from 'lit-element';
import { ArcOverlayMixin } from '@advanced-rest-client/arc-overlay-mixin';

/**
 * Material design: [Bottom sheets](https://material.google.com/components/bottom-sheets.html#)
 *
 * # `<bottom-sheet>`
 *
 * Bottom sheets slide up from the bottom of the screen to reveal more content.
 *
 * ### Example
 *
 * ```html
 * <bottom-sheet>
 *    <paper-icon-item>
 *      <iron-icon src="inbox.png" item-icon></iron-icon>
 *      Inbox
 *    </paper-icon-item>
 *    <paper-icon-item>
 *      <iron-icon src="keep.png" item-icon></iron-icon>
 *      Keep
 *    </paper-icon-item>
 *    <paper-icon-item>
 *      <iron-icon src="hangouts.png" item-icon></iron-icon>
 *      Hangouts
 *    </paper-icon-item>
 *  </bottom-sheet>
 * ```
 *
 * ### Positioning
 *
 * Use the `fit-bottom` class to position the bar at the bottom of the app and with full width;
 *
 * Use `center-bottom` class to display the bar at the bottom centered on a page.
 *
 * ### Styling
 *
 * `<bottom-sheet>` provides the following custom properties and mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--bottom-sheet-background-color` | The bottom-sheet background-color | `#fff`
 * `--bottom-sheet-color` | The bottom-sheet color | `#323232`
 * `--bottom-sheet-max-width` | Max width of the element | ``
 * `--bottom-sheet-max-height` | Max height of the element | ``
 * `--bottom-sheet-label-color` | Color of the label | `rgba(0, 0, 0, 0.54)`
 * `--bottom-sheet-box-shadow` | Box shadow property of the element | `0 2px 5px 0 rgba(0, 0, 0, 0.26)`
 *
 */
export class BottomSheetElement extends ArcOverlayMixin(LitElement) {
  static readonly styles: CSSResult;

  render(): TemplateResult;
  /**
   * The label of the bottom sheet.
   * @attribute
   */
  label: string;
  /**
   * Removes padding from the element styles
   * @attribute
   */
  noPadding: boolean;

  fitInto: HTMLElement;

  /**
   * Returns the scrolling element.
   */
  readonly scrollTarget: HTMLElement;

  constructor();

  connectedCallback(): void;

  disconnectedCallback(): void;

  firstUpdated(): void;

  _openedChanged(opened: boolean): void;

  /**
   * Overridden from `ArcOverlayMixin`.
   */
  _renderOpened(): void;

  /**
   * Overridden from `ArcOverlayMixin`.
   */
  _renderClosed(): void;

  _onFitIntoChanged(fitInto: HTMLElement): void;
}


src/BottomSheetElement.js
/**
@license
Copyright 2018 The Advanced REST Client authors
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html } from 'lit-element';
import { ArcOverlayMixin } from '@advanced-rest-client/arc-overlay-mixin';
import { IronA11yAnnouncer } from '@polymer/iron-a11y-announcer';
import sheetStyles from './styles.js';

/** @typedef {import('lit-html').TemplateResult} TemplateResult */

// Keeps track of the toast currently opened.
let currentSheet;
/**
 * Material design: [Bottom sheets](https://material.google.com/components/bottom-sheets.html#)
 *
 * # `<bottom-sheet>`
 *
 * Bottom sheets slide up from the bottom of the screen to reveal more content.
 *
 * ### Example
 *
 * ```html
 * <bottom-sheet>
 *    <paper-icon-item>
 *      <iron-icon src="inbox.png" item-icon></iron-icon>
 *      Inbox
 *    </paper-icon-item>
 *    <paper-icon-item>
 *      <iron-icon src="keep.png" item-icon></iron-icon>
 *      Keep
 *    </paper-icon-item>
 *    <paper-icon-item>
 *      <iron-icon src="hangouts.png" item-icon></iron-icon>
 *      Hangouts
 *    </paper-icon-item>
 *  </bottom-sheet>
 * ```
 *
 * ### Positioning
 *
 * Use the `fit-bottom` class to position the bar at the bottom of the app and with full width;
 *
 * Use `center-bottom` class to display the bar at the bottom centered on a page.
 *
 * ### Styling
 *
 * `<bottom-sheet>` provides the following custom properties and mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--bottom-sheet-background-color` | The bottom-sheet background-color | `#fff`
 * `--bottom-sheet-color` | The bottom-sheet color | `#323232`
 * `--bottom-sheet-max-width` | Max width of the element | ``
 * `--bottom-sheet-max-height` | Max height of the element | ``
 * `--bottom-sheet-label-color` | Color of the label | `rgba(0, 0, 0, 0.54)`
 * `--bottom-sheet-box-shadow` | Box shadow property of the element | `0 2px 5px 0 rgba(0, 0, 0, 0.26)`
 *
 */
export class BottomSheetElement extends ArcOverlayMixin(LitElement) {
  static get styles() {
    return sheetStyles;
  }

  render() {
    const { label } = this;
    return html`
    ${label ? html`<label>${label}</label>` : ''}
    <div class="scrollable">
      <slot></slot>
    </div>`;
  }

  static get properties() {
    return {
      /**
       * The label of the bottom sheet.
       * @attribute
       */
      label: { type: String },
      /**
       * Removes padding from the element styles
       * @attribute
       */
      noPadding: { type: Boolean, reflect: true }
    };
  }

  get fitInto() {
    return this._fitInto;
  }

  set fitInto(value) {
    const old = this._fitInto;
    if (old === value) {
      return;
    }
    this._fitInto = value;
    this._onFitIntoChanged(value);
  }

  /**
   * Returns the scrolling element.
   * @return {HTMLElement}
   */
  get scrollTarget() {
    return this.shadowRoot.querySelector('.scrollable');
  }

  constructor() {
    super();
    this.__onTransitionEnd = this.__onTransitionEnd.bind(this);

    this.fitInto = window;
    this.opened = false;
    this.label = undefined;
  }

  connectedCallback() {
    /* istanbul ignore else  */
    if (super.connectedCallback) {
      super.connectedCallback();
    }
    // @ts-ignore
    IronA11yAnnouncer.requestAvailability();
    this.addEventListener('transitionend', this.__onTransitionEnd);
  }

  disconnectedCallback() {
    /* istanbul ignore else  */
    if (super.disconnectedCallback) {
      super.disconnectedCallback();
    }
    this.removeEventListener('transitionend', this.__onTransitionEnd);
  }

  firstUpdated() {
    this.sizingTarget = this.scrollTarget;
  }

  _openedChanged(opened) {
    if (opened) {
      if (currentSheet && currentSheet !== this) {
        currentSheet.close();
      }
      currentSheet = this;
      this.dispatchEvent(new CustomEvent('iron-announce', {
        bubbles: true,
        composed: true,
        detail: {
          text: 'Menu opened'
        }
      }));
    } else if (currentSheet === this) {
      currentSheet = null;
    }
    super._openedChanged(opened);
  }

  /**
   * Overridden from `ArcOverlayMixin`.
   */
  _renderOpened() {
    const node = this;
    node.classList.add('bottom-sheet-open');
  }

  /**
   * Overridden from `ArcOverlayMixin`.
   */
  _renderClosed() {
    const node = this;
    node.classList.remove('bottom-sheet-open');
  }

  /**
   * @private
   * @param {HTMLElement} fitInto
   */
  _onFitIntoChanged(fitInto) {
    this.positionTarget = fitInto;
  }

  __onTransitionEnd(e) {
    // there are different transitions that are happening when opening and
    // closing the toast. The last one so far is for `opacity`.
    // This marks the end of the transition, so we check for this to determine if this
    // is the correct event.
    if (e && e.target === this && e.propertyName === 'opacity') {
      if (this.opened) {
        this._finishRenderOpened();
      } else {
        this._finishRenderClosed();
      }
    }
  }
  /**
   * Fired when `bottom-sheet` is opened.
   *
   * @event 'iron-announce'
   * @param {{text: string}} detail Contains text that will be announced.
   */
}
